python/node_similarity.py

```diff
@@ -0,0 +1,302 @@
+
+    if denominator != 0:
+        cosine_similarity = len(neighbors1 & neighbors2) / denominator
+
+    return cosine_similarity
+
+
+def _calculate_similarity(
+    node1: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    node2: Union[mgp.Vertex, Tuple[mgp.Vertex]],
+    method: Callable,
+    mode: str,
+) -> List[Tuple[mgp.Vertex, mgp.Vertex, float]]:
+    """This method calculates the similarity of nodes for given method and mode.
+
+    :param node1: The first node or tuple of nodes
+    :type node1: mgp.Vertex, tuple
+    :param node2: The second node or tuple of nodes
+    :type node2: mgp.Vertex, tuple
+    :param method: Similarity measure which will be used for calculating the similarity between nodes.
+    Currently available are `_calculate_jaccard`, `_calculate_overlap` and `_calculate_cosine`
+    :type method: function
+    :param mode: Can be `p` (or `pairwise`) for pairwise similarity calculation or `c` (or `cartesian`) for
+    calculating the similarity of the Cartesian product of given tuples
+    :type mode: str
+
+    :raises TypeError: Occurs if there's a type mismatch for passed arguments
+    :raises ValueError: Occurs if any passed argument is invalid
+
+    :return: Returns the calculated similarity between nodes in a list of mgp.Records. Each mgp.Record consist of
+    the first node, the second node and the similarity between them.
+    :rtype: list
+    """
+
+    nodes1: tuple
+    nodes2: tuple
+
+    if isinstance(node1, mgp.Vertex):
+        nodes1 = tuple([node1])
+    elif isinstance(node1, tuple):
+        nodes1 = node1
+    else:
+        raise TypeError("Invalid type of first argument.")
+
+    if isinstance(node2, mgp.Vertex):
+        nodes2 = tuple([node2])
+    elif isinstance(node2, tuple):
+        nodes2 = node2
+    else:
+        raise TypeError("Invalid type of second argument.")
+
+    mode_constants = Mode()
+
+    result: list
+
+    if mode.lower() in mode_constants.PAIRWISE:
+        if len(nodes1) == len(nodes2):
+            return [
+                mgp.Record(node1=n1, node2=n2, similarity=method(n1, n2))
+                for n1, n2 in zip(nodes1, nodes2)
+            ]
+        else:
+            raise ValueError("Incompatible lengths of given arguments.")
+    elif mode.lower() in mode_constants.CARTESIAN:
+        return [
+            mgp.Record(node1=n1, node2=n2, similarity=method(n1, n2))
+            for n1, n2 in product(nodes1, nodes2)
+        ]
+    else:
+        raise ValueError("Invalid mode.")
+
+
+def _get_neighbors(node: mgp.Vertex) -> Set[mgp.Vertex]:
+    """This method finds all neighbors of a given node.
+
+    :param node: A node
+    :type node: mgp.Vertex
+
+    :return: Set of all neighbors of a node
+    :rtype: set
+    """
+
+    neighbors = set()
+
+    for edge in node.in_edges:
+        neighbors.add(edge.from_vertex)
+
+    for edge in node.out_edges:
+        neighbors.add(edge.to_vertex)
+
+    neighbors_dict[node.id] = neighbors
```

Comment:
Add a comment that this is used to preserve state. Also, this needs to be updated! Be careful, when calling the module again, after creating more nodes, this code would return the previous state, because as any other global static variable, this one is living in the global scope once is created.